File: scala/src/main/scala/part1/chapter4/Validated.scala
package part1.chapter4

import scala.{Either, Left, Right}

case class Name private (value: String)
object Name:
    def apply(name: String): Either[String, Name] =
        if name == "" || name == null then Left("Name is empty.")
        else Right(new Name(name))

case class Age private (value: Int)
object Age:
    def apply(age: Int): Either[String, Age] =
        if age < 0 then Left("Age is out of range.")
        else Right(new Age(age))

case class Person(name: Name, age: Age)
object Person:
    def makeBoth(name: String, age: Int): Either[List[String], Person] =
        map2Both(Name(name), Age(age), Person(_, _))

def map2Both[E, A, B, C](a: Either[E, A], b: Either[E, B], f: (A, B) => C): Either[List[E], C] =
    (a, b) match
        case (Right(aa), Right(bb)) => Right(f(aa, bb))
        case (Left(e), Right(_)) => Left(List(e))
        case (Right(_), Left(e)) => Left(List(e))
        case (Left(e1), Left(e2)) => Left(List(e1, e2))

def map2All[E, A, B, C](a: Either[List[E], A], b: Either[List[E], B], f: (A, B) => C): Either[List[E], C] =
    (a, b) match
        case (Right(aa), Right(bb)) => Right(f(aa, bb))
        case (Left(es), Right(_)) => Left(es)
        case (Right(_), Left(es)) => Left(es)
        case (Left(es1), Left(es2)) => Left(es1 ++ es2)

def traverseAll[E, A, B](as: List[A], f: A => Either[List[E], B]): Either[List[E], List[B]] =
    as.foldRight(Right(Nil): Either[List[E], List[B]])((a, acc) => map2All(f(a), acc, _ :: _))

enum Validated[+E, +A]:
    case Invalid(errors: List[E])
    case Valid(get: A)

    def toEither: Either[List[E], A] =
        this match
            case Valid(a) => Right(a)
            case Invalid(es) => Left(es)

    def map[B](f: A => B): Validated[E, B] =
        this match
            case Valid(a) => Valid(f(a))
            case Invalid(es) => Invalid(es)

    def map2[EE >: E, B, C](b: Validated[EE, B])(f: (A, B) => C): Validated[EE, C] =
        (this, b) match
            case (Valid(aa), Valid(bb)) => Valid(f(aa, bb))
            case (Invalid(es), Valid(_)) => Invalid(es)
            case (Valid(_), Invalid(es)) => Invalid(es)
            case (Invalid(es1), Invalid(es2)) => Invalid(es1 ++ es2)

object Validated:
    def fromEither[E, A](e: Either[List[E], A]): Validated[E, A] =
        e match
            case Right(a) => Valid(a)
            case Left(es) => Invalid(es)

    def traverse[E, A, B](as: List[A], f: A => Validated[E, B]): Validated[E, List[B]] =
        as.foldRight(Valid(Nil): Validated[E, List[B]])((a, acc) => f(a).map2(acc)(_ :: _))

    def sequence[E, A](vs: List[Validated[E, A]]): Validated[E, List[A]] =
        traverse(vs, identity)

File: rust/src/part1/chapter4/validated.rs
use crate::part1::chapter4::either::Either;
use std::iter::once;

#[derive(Clone)]
enum Validated<E, A> {
    Invalid(Vec<E>),
    Valid(A),
}

impl<E, A> Validated<E, A>
where
    E: Clone,
    A: Clone,
{
    pub fn to_either(&self) -> Either<Vec<E>, A> {
        match self {
            Validated::Invalid(es) => Either::Left(es.clone()),
            Validated::Valid(a) => Either::Right(a.clone()),
        }
    }

    pub fn map<B, F>(&self, f: F) -> Validated<E, B>
    where
        F: Fn(A) -> B + Copy,
    {
        match self {
            Validated::Invalid(es) => Validated::Invalid(es.clone()),
            Validated::Valid(a) => Validated::Valid(f(a.clone())),
        }
    }

    pub fn map2<B, C, F>(&self, b: Validated<E, B>, f: F) -> Validated<E, C>
    where
        F: Fn(A, B) -> C,
    {
        match (self, b) {
            (Validated::Valid(aa), Validated::Valid(bb)) => Validated::Valid(f(aa.clone(), bb)),
            (Validated::Invalid(es), Validated::Valid(_)) => Validated::Invalid(es.clone()),
            (Validated::Valid(_), Validated::Invalid(es)) => Validated::Invalid(es),
            (Validated::Invalid(es1), Validated::Invalid(es2)) => {
                Validated::Invalid(es1.iter().cloned().chain(es2.iter().cloned()).collect())
            }
        }
    }

    pub fn from_either(e: Either<Vec<E>, A>) -> Validated<E, A> {
        match e {
            Either::Right(a) => Validated::Valid(a),
            Either::Left(es) => Validated::Invalid(es),
        }
    }

    fn traverse<B, F>(xs: &Vec<A>, f: F) -> Validated<E, Vec<B>>
    where
        B: Clone,
        F: Fn(A) -> Validated<E, B> + Copy,
    {
        xs.iter().fold(Validated::Valid(Vec::new()), |acc, a| {
            f(a.clone()).map2(acc, |head, tail| once(head).chain(tail).collect())
        })
    }

    fn sequence(xs: &Vec<Validated<E, A>>) -> Validated<E, Vec<A>> {
        Validated::traverse(&xs, |x| x)
    }
}
